feat(experiments): bring MDE setting back

jurajmajerik opened the pull request:
## Problem
Prompted by user feedback:
> I no longer see the step when creating an experiment to define the minimum acceptable improvement (minimum detectable effect). Did this move somewhere new with the new UI changes?

> Using a 4% absolute MDE that you're calculating here seems a bit wild to me considering that the baseline conversation rate is somewhere around 18%.

> This should either be a much lower absolute MDE or even better we should be able to set it to use a relative value because that's ultimately how our PMs and TPMs think


## Changes
- Add back the ability to set Minimum Detectable Effect (MDE)
- Set MDE back to the recommended setting after a goal has been updated

<img width="623" alt="image" src="https://github.com/PostHog/posthog/assets/22996112/149a8c52-8496-4464-892b-1e0b67349851">

<img width="1016" alt="image" src="https://github.com/PostHog/posthog/assets/22996112/7df1118e-bda1-4657-a734-087986e798be">

## How did you test this code?
👀

jurajmajerik committed:
init

jurajmajerik committed:
init

jurajmajerik committed:
clean up

jurajmajerik committed:
clean up

jurajmajerik committed:
reset MDE after goal updated

jurajmajerik committed:
adjust style

jurajmajerik committed:
clean up

jurajmajerik committed:
tweaks

jurajmajerik committed:
clean up

neilkakkar commented on frontend/src/scenes/experiments/ExperimentView/DataCollectionCalculator.tsx:
Probably a  backwards compatibility issue: With existing experiments, with no MDE set, the icon hover does nothing and then on clicking recalculate it shoots upto 100
<img width="494" alt="image" src="https://github.com/PostHog/posthog/assets/7115141/ffebc1a1-cce1-46b6-845a-7bdf02b0efd1">

<img width="470" alt="image" src="https://github.com/PostHog/posthog/assets/7115141/218999b9-f68e-481b-a204-aa885b56ee35">

which I suppose is default when null? Not sure if possible to reproduce with new experiments, or if existing users will run into this, but can you double check?

One quick way to check is check in metabase if exp parameters don't have mde, and how many such experiments are there.

jurajmajerik replied:
1.
> with no MDE set, the icon hover does nothing
Fixed - when the MDE value has not been persisted on the record, let's not show the tooltip.

2.
> and then on clicking recalculate it shoots upto 100
This is because the baseline count is too small: https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/experiments/utils.ts#L82
So not a backwards compatibility issue. Of course, the slider max value should adapt to the current value which doesn't happen here - will fix.

3.
> check in metabase if exp parameters don't have mde
Haven't checked, but basically any experiment where the creator didn't touch the MDE slider will not have it set.
When MDE is not saved on the record, we don't know what value was used to calculate the current estimation (although technically we can infer the default of 20%?), and so we simply shouldn't show the tooltip (see the first point). For newly created experiments this value will always be persisted and available.

neilkakkar commented on frontend/src/scenes/experiments/ExperimentView/DataCollectionCalculator.tsx:
oh seems like it only happens on trends

jurajmajerik replied:
hmm this seems like an insight loading issue & non-deterministic, let me see 👀

neilkakkar replied:
I could reliably repro on trends, but maybe because no data in last 14 days? 🤔

jurajmajerik replied:
this was a me issue after all 🤦‍♂️ fixed!

neilkakkar commented on frontend/src/scenes/experiments/ExperimentView/DataCollectionCalculator.tsx:
```diff
@@ -0,0 +1,267 @@
+                            )}
+                        </div>
+                    )}
+                    <div className="flex w-full mt-4">
```
I'd say none of this data is necessary here anymore, all I want to see is the slider, and how it changes the recommended running time / sample size.

neilkakkar replied:
<img width="1002" alt="image" src="https://github.com/PostHog/posthog/assets/7115141/13897c91-25ce-4232-a4d4-2fdf1b3d4d37">

Feels the rest is extra out of context

neilkakkar reviewed (commented):
Nice, thanks for bringing this back!

jurajmajerik committed:
remove redundant parts

jurajmajerik committed:
fix margins

jurajmajerik committed:
use single tooltip for MDE & high running time

jurajmajerik commented on frontend/src/scenes/experiments/ExperimentView/DataCollectionCalculator.tsx:
good catch, merged into a single tooltip:

![image](https://github.com/PostHog/posthog/assets/22996112/f44a10cb-0209-4776-b70b-c6d7408ba0bb)

jurajmajerik committed:
fix widths

jurajmajerik committed:
fix missing trendCount

jurajmajerik committed:
refactor funnel/trend calculations

jurajmajerik committed:
add info banner

jurajmajerik committed:
add disabled reason

jurajmajerik committed:
fix Funnel recommended running time

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiment-mde

jurajmajerik commented:
Changes:
- Remove unnecessary parts from the modal (variants etc.)
- Use a single tooltip for the MDE & high running time warning
- Fix missing recommended running time for Trends
- Adjust MDE slider max value for Trends (up to 100%)
- Don't show the tooltip if the MDE is not persisted (old records only)
- Fixed a long-standing bug when a Funnel calculator showed a Trend experiment running time.
- Some refactoring - split Funnel and Trend calculation UI into separate components, adjusted naming, moved some calculations into the logic. I want to spend more time tidying up/streamlining various calculations and naming, but this is worth being a separate sprint goal.
- Add an info banner explaining why the currently calculated running time may differ from previous estimates

<img width="553" alt="image" src="https://github.com/PostHog/posthog/assets/22996112/bcc3a358-387f-4d64-900f-19661fefbeb8">

jurajmajerik commented:
[note to self]
Before merging:
- rename ProgressBar.tsx -> DataCollection.tsx
- rename minimumDetectableChange -> minimumDetectableEffect

jurajmajerik commented:
One remaining issue:
- open the calculator
- set new MDE
- close the calculator without saving
- the MDE tooltip (next to the goal) now shows the new MDE, even though the changes have not been saved

This is because the calculator state and the record state are intermingled. I want to solve this in the follow-up refactor by having a pure functions for calculating the statistical values. For now I suggest hiding the tooltip if the MDE value has been touched 😬

neilkakkar reviewed (approved):
Didn't run locally, but looks good! :face_with_peeking_eye:

neilkakkar commented:
>  For now I suggest hiding the tooltip if the MDE value has been touched 😬

Hmmm, this isn't ideal. Sounds like the problem here is that as soon as the mde changes we save it on the local experiment. Which I guess is ok, that's how everything else does this.

To combat this problem, what everything else does is have a touch handler: https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/experiments/experimentLogic.tsx#L301 - you can add to this reducer and it should automatically handle reloads whenever its touched but not saved.

jurajmajerik committed:
reload experiment after closing the modal

jurajmajerik committed:
fix names

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiment-mde
